### DISCUSSION TITLE
"UNIQUE constraint failed: users.tokenkey" when create user from custom route

### DISCUSSION THREAD

**Discussion by demirdev** [pocketbase/pocketbase#1682] [Q&A]
How can I solve this problem?
Thanks a lot.

<img width="1373" alt="Screen Shot 2023-01-25 at 12 41 20" src="https://user-images.githubusercontent.com/54906407/214529964-fb5d9f47-964f-4bcd-9a19-4351bab2b9ec.png">

**Comment by ganigeorgiev** [comment]
How do you create the user record? Do you explicitly set a `TokenKey` or it is empty (if it is empty that could explain the unique constraint error)?

I'm not sure that I understand the above screenshot but if you need to set a password for an auth record, the correct way is to use the `record.SetPassword(plainPassword)` method since it will properly hash the plain password with bcrypt and will always refresh the internal TokenKey.

List of the available auth record helpers you can find in https://pocketbase.io/docs/record-methods/#getset-record-fields.

**Reply by demirdev** [reply]
Thank you @ganigeorgiev , I added this to solve it. 
Best regards.
<img width="829" alt="Screen Shot 2023-01-25 at 12 53 05" src="https://user-images.githubusercontent.com/54906407/214532656-5314c8c0-9433-4cf5-8b08-3a2ed5f9c614.png">

**Reply by ganigeorgiev** [reply]
@demirdev Please note that the above code is not entirely correct (there is no "passwordConfirm" field in the user record).

Since you are using already the `forms.RecordUpsert` you can just leave the persistent to it, something like:


```go
user := models.NewRecord(usersCollection)
password := security.RandomString(70)

form := forms.NewRecordUpsert(app, user)
form.SetFullManageAccess(true)
form.SetDao(txDao)

form.LoadData(map[string]any{
    "username": username,
    "password": password,
    "passwordConfirm": password,
})

if err := form.Submit(); err != nil {
    return apis.NewBadRequestError("Failed to create user", err)
}
```

**Comment by demirdev** [comment]
I found that from your reply: form.setDao(txDao). I will use it and then form.Submit.

My scenario is like this, if I can do it, I will do it.

I am making a simple accounting module with collections like products, sales, purchases, balance using Pocketbase.

I will add it as a package to 3 mobile apps written using dart/flutter.  

When these mobile app users arrive, they will send their idToken from firebase authentication.

I verify them and if necessary, I open their accounts in pocketbase as above. But in this usage, users will not enter a password. They will always log in with idToken verification.

My intention was not to use a password, but when it gave an error, I decided to give a random password with long characters.

Can you please suggest something for me if you can?

Thank you very much.

**Reply by ganigeorgiev** [reply]
I don't have enough context for the use case to suggest an alternative.

The above example form submission is fine if you want to create a user with data validations.
The password field is always required so setting a random password (_even 20-30 characters should be enough_) is the recommended way when you don't know/care what is the user password during the creation (we use the same approach when creating a new auth record during the OAuth2 sign-up process).
